lib/plugins/bull-mq-metrics/MetricsCollector.ts

```diff
@@ -0,0 +1,176 @@
+    buckets: histogramBuckets,
+    labelNames: ['queue'],
+  }),
+})
+
+export class MetricsCollector {
+  private readonly options: MetricCollectorOptions
+  private readonly metrics: Metrics
+  private active = true
+
+  constructor(
+    private readonly redis: Redis,
+    private readonly registry: prometheus.Registry,
+    private readonly logger: FastifyBaseLogger,
+    options: Partial<MetricCollectorOptions>,
+  ) {
+    this.options = {
+      bullMqPrefix: 'bull',
+      metricsPrefix: 'bullmq',
+      excludedQueues: [],
+      collectionIntervalInMs: 5000,
+      histogramBuckets: [20, 50, 150, 400, 1000, 3000, 8000, 22000, 60000, 150000],
+      queueDiscoverer: new BackgroundJobsBasedQueueDiscoverer(this.redis),
+      ...options,
+    }
+
+    this.metrics = this.registerMetrics(this.registry, this.options)
+  }
+
+  async start() {
+    const queueNames = await this.options.queueDiscoverer.discoverQueues()
+    await Promise.all(
+      queueNames.map((name) => this.observeQueue(name, this.redis, this.metrics, this.options)),
+    )
+  }
+
+  stop() {
+    this.active = false
+  }
+
+  private registerMetrics(
+    registry: prometheus.Registry,
+    { metricsPrefix, histogramBuckets }: MetricCollectorOptions,
+  ): Metrics {
+    const metrics = getMetrics(metricsPrefix, histogramBuckets)
+    const metricNames = Object.keys(metrics)
+
+    // If metrics are already registered, just return them to avoid triggering a Prometheus error
+    if (metricNames.length > 0 && registry.getSingleMetric(metricNames[0])) {
+      const retrievedMetrics = registry.getMetricsAsArray()
+      const returnValue: Record<string, prometheus.MetricObject> = {}
+
+      for (const metric of retrievedMetrics) {
+        if (metricNames.includes(metric.name)) {
+          returnValue[metric.name as keyof Metrics] = metric
+        }
+      }
+
+      return returnValue as unknown as Metrics
+    }
+
+    for (const metric of Object.values(metrics)) {
+      registry.registerMetric(metric)
+    }
+
+    return metrics
+  }
+
+  private async observeQueue(
+    name: string,
+    redis: Redis,
+    metrics: Metrics,
+    { collectionIntervalInMs }: MetricCollectorOptions,
+  ) {
+    const queue = new Queue(name, { connection: redis })
+    const events = new QueueEvents(name, { connection: redis })
+
+    events.on('completed', (completedJob: { jobId: string }) => {
+      queue
+        .getJob(completedJob.jobId)
+        .then((job) => {
+          if (!job) {
+            return
+          }
+
+          if (job.finishedOn) {
+            metrics.completedDuration
+              .labels({ queue: name })
+              .observe(job.finishedOn - job.timestamp)
+
+            if (job.processedOn) {
+              metrics.processedDuration
+                .labels({ queue: name })
+                .observe(job.finishedOn - job.processedOn)
+            }
+          }
+        })
+        .catch((err) => {
+          this.logger.warn(err)
```

Comment:
why we consider this a warning and not an error?

Reply:
I thought about warning because this is non critical for the app execution - this is just monitoring giving out flukes (we don't really expect errors here).

I can change to `error` if you want.

Reply:
no strong opinion, either one works